### PR TITLE
gnss: API for retrieving PPS timestamp

### PR DESCRIPTION
**Pull request by JordanYates**
Add an API function for retrieving the timestamp of the latest PPS timepulse at the highest resolution we have available, the kernel tick count. Added the timepulse information to the GNSS sample.

In the true spirit of the GNSS API, the timepulse pin behaviour is "Whatever the modem does by default" :)

I don't have any hardware with in-tree drivers, but the updated sample works with my OOT driver with a PPS GPIO.

<details><summary>Output log at transition from no PPS to PPS</summary>
<p>

```
Got a fix!
13 satellites reported (of which 7 tracked)!
sam_m10@42: gnss_info: {satellites_cnt: 6, hdop: 4.160, fix_status: GNSS_FIX, fix_quality: GNSS_SPS}
sam_m10@42: navigation_data: {latitude: ********, longitude : **********, bearing 0.000, speed 0.047, altitude: 52.908}
sam_m10@42: gnss_time: {hour: 1, minute: 6, millisecond 39000, month_day 30, month: 7, century_year: 24}
sam_m10@42: gnss_satellite: {prn: 8, snr: 30, elevation 15, azimuth 139, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 13, snr: 27, elevation 165, azimuth 0, system: GPS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 14, snr: 35, elevation 61, azimuth 233, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 30, snr: 43, elevation 63, azimuth 169, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 128, snr: 0, elevation 9, azimuth 280, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 129, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 133, snr: 0, elevation 2, azimuth 84, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 137, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 6, snr: 36, elevation 165, azimuth 0, system: GALILEO, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 2, snr: 38, elevation 60, azimuth 291, system: QZSS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 14, snr: 34, elevation 28, azimuth 138, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 15, snr: 40, elevation 42, azimuth 203, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 255, snr: 36, elevation 165, azimuth 0, system: GLONASS, is_tracked: 0}
Got a fix @ 48705 ms
13 satellites reported (of which 7 tracked)!
sam_m10@42: gnss_info: {satellites_cnt: 6, hdop: 4.160, fix_status: GNSS_FIX, fix_quality: GNSS_SPS}
sam_m10@42: navigation_data: {latitude: **********, longitude : **********, bearing 0.000, speed 0.029, altitude: 53.044}
sam_m10@42: gnss_time: {hour: 1, minute: 6, millisecond 40000, month_day 30, month: 7, century_year: 24}
sam_m10@42: gnss_satellite: {prn: 8, snr: 30, elevation 15, azimuth 139, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 13, snr: 27, elevation 165, azimuth 0, system: GPS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 14, snr: 35, elevation 61, azimuth 233, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 30, snr: 43, elevation 63, azimuth 169, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 128, snr: 0, elevation 9, azimuth 280, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 129, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 133, snr: 0, elevation 2, azimuth 84, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 137, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 6, snr: 35, elevation 165, azimuth 0, system: GALILEO, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 2, snr: 38, elevation 60, azimuth 291, system: QZSS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 14, snr: 34, elevation 28, azimuth 138, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 15, snr: 40, elevation 42, azimuth 203, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 255, snr: 36, elevation 165, azimuth 0, system: GLONASS, is_tracked: 0}
Got a fix @ 49705 ms
13 satellites reported (of which 7 tracked)!
sam_m10@42: gnss_info: {satellites_cnt: 6, hdop: 4.160, fix_status: GNSS_FIX, fix_quality: GNSS_SPS}
sam_m10@42: navigation_data: {latitude: **********, longitude : ***********, bearing 0.000, speed 0.022, altitude: 53.191}
sam_m10@42: gnss_time: {hour: 1, minute: 6, millisecond 41000, month_day 30, month: 7, century_year: 24}
sam_m10@42: gnss_satellite: {prn: 8, snr: 30, elevation 15, azimuth 139, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 13, snr: 26, elevation 165, azimuth 0, system: GPS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 14, snr: 34, elevation 61, azimuth 233, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 30, snr: 43, elevation 63, azimuth 169, system: GPS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 128, snr: 0, elevation 9, azimuth 280, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 129, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 133, snr: 0, elevation 2, azimuth 84, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 137, snr: 0, elevation 47, azimuth 313, system: SBAS, is_tracked: 0}
sam_m10@42: gnss_satellite: {prn: 6, snr: 35, elevation 165, azimuth 0, system: GALILEO, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 2, snr: 39, elevation 60, azimuth 291, system: QZSS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 14, snr: 34, elevation 28, azimuth 138, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 15, snr: 40, elevation 42, azimuth 203, system: GLONASS, is_tracked: 1}
sam_m10@42: gnss_satellite: {prn: 255, snr: 37, elevation 165, azimuth 0, system: GLONASS, is_tracked: 0}

```

</p>
</details> 

Example driver implementation:
```
static int ubx_m10_i2c_get_latest_timepulse(const struct device *dev, k_ticks_t *timestamp)
{
	const struct ubx_m10_i2c_config *cfg = dev->config;
	struct ubx_m10_i2c_data *data = dev->data;

	if (cfg->timepulse_gpio.port == NULL) {
		/* No timepulse pin connected */
		return -ENOTSUP;
	}
	if (data->latest_timepulse == 0) {
		/* Timepulse interrupt has not occured yet */
		return -EAGAIN;
	}
	*timestamp = data->latest_timepulse;
	return 0;
}

static void timepulse_gpio_callback(const struct device *dev, struct gpio_callback *cb,
				    uint32_t pins)
{
	struct ubx_m10_i2c_data *data = CONTAINER_OF(cb, struct ubx_m10_i2c_data, timepulse_cb);

	data->latest_timepulse = k_uptime_ticks();
	LOG_DBG("");
}
```

### CODE REVIEW

File: include/zephyr/drivers/gnss.h [hunk 1]
```diff
@@ -451,6 +456,33 @@ static inline int z_impl_gnss_get_supported_systems(const struct device *dev,
 	return api->get_supported_systems(dev, systems);
 }
 
+/**
+ * @brief Get the timestamp of the latest PPS timepulse
+ *
+ * @note PPS timepulse is considered valid when actively toggling
```
Comment (tomi-font) [inline include/zephyr/drivers/gnss.h:462]: Actively toggling what?

File: samples/drivers/gnss/src/main.c [hunk 2]
```diff
@@ -15,8 +15,16 @@ LOG_MODULE_REGISTER(gnss_sample, CONFIG_GNSS_LOG_LEVEL);
 
 static void gnss_data_cb(const struct device *dev, const struct gnss_data *data)
 {
+	uint64_t timepulse_ms;
+	k_ticks_t timepulse;
+
 	if (data->info.fix_status != GNSS_FIX_STATUS_NO_FIX) {
-		printf("Got a fix!\n");
+		if (gnss_get_latest_timepulse(dev, &timepulse) == 0) {
+			timepulse_ms = k_ticks_to_ms_near64(timepulse);
```
Comment (tomi-font) [inline samples/drivers/gnss/src/main.c:23]: How about using `k_ticks_to_ns_near64()` instead?

File: include/zephyr/drivers/gnss.h (hunk 1 above)
Comment (JordanYates) [reply include/zephyr/drivers/gnss.h:462]: The timepulse pin, it actively toggles itself.
It is valid when it is doing the thing, not valid when it isn't
image

Comment (tomi-font) [reply include/zephyr/drivers/gnss.h:462]: Alright. Let me suggest another phrasing:
```suggestion
 * @note The timestamp is considered valid when the timepulse pin is actively toggling.
```

**Comment by bjarki-andreasen** [comment]
The reason this API does not exist is that if this high of a resolution/synchronization is required, which is typically the case for applications using PPS, the PPS output would typically be connected to a hardware timer or other hardware mechanism, rather than routed through a GPIO callback, which inherently adds a ton of overhead.

**Comment by JordanYates** [comment]
> The reason this API does not exist is that if this high of a resolution/synchronization is required, which is typically the case for applications using PPS, the PPS output would typically be connected to a hardware timer or other hardware mechanism, rather than routed through a GPIO callback, which inherently adds a ton of overhead.

So I have the following logs of this in operation:
```
[00:08:12.796,234] <inf> ublox_m10: TP
[00:08:12.858,642] <inf> task_gnss: Valid: F7 Flags: 01 Flags2: EA Flags3; 0000
[00:08:12.858,642] <inf> task_gnss: LAT: -274914334 LON: 1529885093 iTOW: 193778000
```
Using the timepulse interrupt means any time sync is going to be around 66ms more accurate in this setup, which is a not insignificant improvement. There may be an offset because of the GPIO interrupt, but its much better than waiting for a comms interface.

**Comment by bjarki-andreasen** [comment]
> > The reason this API does not exist is that if this high of a resolution/synchronization is required, which is typically the case for applications using PPS, the PPS output would typically be connected to a hardware timer or other hardware mechanism, rather than routed through a GPIO callback, which inherently adds a ton of overhead.
> 
> So I have the following logs of this in operation:
> 
> ```
> [00:08:12.796,234] <inf> ublox_m10: TP
> [00:08:12.858,642] <inf> task_gnss: Valid: F7 Flags: 01 Flags2: EA Flags3; 0000
> [00:08:12.858,642] <inf> task_gnss: LAT: -27491 LON: 152988 iTOW: 193778000
> ```
> 
> Using the timepulse interrupt means any time sync is going to be around 66ms more accurate in this setup, which is a not insignificant improvement. There may be an offset because of the GPIO interrupt, but its much better than waiting for a comms interface.

I agree it is more accurate, but adding this complexity is just not worth it IMO, the issue with PPS and synchronization is an issue not directly related to GNSS, GNSS just happens to expose a PPS signal, PPS would require a subsystem of its own. A user can also set a higher baudrate to acheive lower delay.

**Comment by bjarki-andreasen** [comment]
> I disagree that GNSS "just happens to expose" a PPS signal, accurate time synchronization is one of the core functionalities of the GNSS networks. Adding an entirely new subsystem to manage a single GPIO that (as far as I know) every GNSS modem exposes doesn't seem like a reasonable solution.

That is not what the subsystem would do... the subsystem would expose an API to get the exact time synchronized by the PPS signal, using the PPS signal as an input to a high resolution timer/counter or other timing hardware (some RTCs can synchronize with and/or use a PPS (1Hz) signal as clock input, see ATSAM4E).

**Comment by JordanYates** [comment]
> That is not what the subsystem would do... the subsystem would expose an API to get the exact time synchronized by the PPS signal, using the PPS signal as an input to a high resolution timer/counter or other timing hardware (some RTCs can synchronize with and/or use a PPS (1Hz) signal as clock input, see ATSAM4E).

That is all well and good, but who is likely to actually implement such a subsystem and in what time frame?
And the fallback case for such a system (for designs without such fancy synchronisation mechanisms) can only ever be what I am proposing here, exposing the kernel tick that the PPS pulse was observed at.

I would be more than happy for this functionality to transition out of GNSS and into the hypothetical time sync subsystem when it exists, the GNSS API is marked as experimental so there is not significant overhead for doing so. I also disagree that the complexity of having this in the GNSS driver is too high, an entire implementation is ~20 lines of bog standard GPIO interrupts, and could be moved into a common file if desired as I don't think there is anything driver specific required.

**Comment by bjarki-andreasen** [comment]
> an entire implementation is ~20 lines of bog standard GPIO interrupts, and could be moved into a common file if desired as I don't think there is anything driver specific required.

Perfect, that is one way an application is expected to use PPS, in which case, why have it be tied to the GNSS API? It is just monitoring a GPIO in this case, all the GNSS can do is configure/enable PPS, how the output is used is not relevant to the GNSS itself.

**Comment by JordanYates** [comment]
> Perfect, that is one way an application is expected to use PPS, in which case, why have it be tied to the GNSS API? It is just monitoring a GPIO in this case, all the GNSS can do is configure/enable PPS, how the output is used is not relevant to the GNSS itself.

I think the fact that the GNSS is configuring and enabling the PPS is enough of a reason to have the output come from the GNSS API as well. The PPS GPIO will certainly be a property on the GNSS devicetree node.

Not doing so is like proposing that the RTC API has all the functions for configuring and enabling an alarm at a certain point in time, but saying that actually handling the alarm is out of scope.

**Comment by bjarki-andreasen** [comment]
> > Perfect, that is one way an application is expected to use PPS, in which case, why have it be tied to the GNSS API? It is just monitoring a GPIO in this case, all the GNSS can do is configure/enable PPS, how the output is used is not relevant to the GNSS itself.
> 
> I think the fact that the GNSS is configuring and enabling the PPS is enough of a reason to have the output come from the GNSS API as well. The PPS GPIO will certainly be a property on the GNSS devicetree node.
> 
> Not doing so is like proposing that the RTC API has all the functions for configuring and enabling an alarm at a certain point in time, but saying that actually handling the alarm is out of scope.

I see the GNSS 1Hz output similar to a configurable clock output of an RTC like https://github.com/zephyrproject-rtos/zephyr/blob/2d2ce4e78d004fc13abbf09e73a36e68b147c365/dts/bindings/rtc/nxp%2Cpcf8523.yaml#L27-L42 which is configured by the RTC device driver. Its an output that can be used for anything, and could be used interchangeably with the output from a GNSS if configured to 1Hz output.

An application could have some dependency on a 1Hz GPIO routed interrupt which is synced with the real-time base, which could be the GNSS, or an RTC if no GNSS is present. Would you argue that the RTC API should have a similar API for its clock output that is being proposed for the GNSS API?

**Comment by JordanYates** [comment]
> An application could have some dependency on a 1Hz GPIO routed interrupt which is synced with the real-time base, which could be the GNSS, or an RTC. Would you argue that the RTC API should have a similar API for its clock output that is being proposed for the GNSS API?

I am unsure what exactly you are arguing here. A clock output is not a core part of an RTC as far as I know.

It also appears we aren't going to come to an agreement on whether this makes sense to be part of the GNSS API, perhaps others following along can chime in? @kartben @tomi-font

**Comment by tomi-font** [comment]
I'm not really familiar with PPS, RTC and how it's like in Zephyr but I think it makes sense to not gatekeep the relatively simple change that @JordanYates is proposing based on the premise that it could be its own bigger thing.
PPS being a functionality of (a certain part of the) GNSS modems, it's not unnatural to have it here, even though PPS could be a more generic feature of its own.
Such additions as proposed here can always be reworked and moved elsewhere, especially since the GNSS API is rather new and marked experimental.
Now, I'm not able to tell whether this should be part of RTC. If the same kind of change is possible and preferred there, then it's fine, but other than that let's not make perfect the enemy of good. Small incremental improvements are better than nothing.

**Comment by henrikbrixandersen** [comment]
> I see the GNSS 1Hz output similar to a configurable clock output of an RTC like
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/2d2ce4e78d004fc13abbf09e73a36e68b147c365/dts/bindings/rtc/nxp%2Cpcf8523.yaml#L27-L42
> which is configured by the RTC device driver. Its an output that can be used for anything, and could be used interchangeably with the output from a GNSS if configured to 1Hz output.
> 
> An application could have some dependency on a 1Hz GPIO routed interrupt which is synced with the real-time base, which could be the GNSS, or an RTC. Would you argue that the RTC API should have a similar API for its clock output that is being proposed for the GNSS API?
> 
> note: The RTC API does have an API for a callback every second called `rtc_update_set_callback()` since this is provided by some integrated RTC devices through an interrupt. It is not related to a potential configurable clock output :)

I tend to agree. The solution proposed here (getting a timestamp of the last pulse) is one way of using a 1PPS signal, but certainly not the only way. I think the complexity of this is low is enough to leave it for the application to capture a timestamp on the 1PPS interrupt, if that is what it needs.

**Comment by JordanYates** [comment]
> I tend to agree. The solution proposed here (getting a timestamp of the last pulse) is one way of using a 1PPS signal, but certainly not the only way. I think the complexity of this is low is enough to leave it for the application to capture a timestamp on the 1PPS interrupt, if that is what it needs.

Unfortunately the application is not really in a great position to use the 1PPS signal for a few reasons.
Firstly the PPS signal can stop when the modem is put into sleep mode. The modem driver obviously knows when this happens and can return the documented error, not necessarily true if the application is handling this.
Secondly, some modems have weird transient behaviors on the PPS signal at power mode transitions. The application layer cannot possibly handle this in a generic way, while a driver can perform whatever masking operations are necessary to provide only valid values to the application.
Thirdly, if delegated to the application, how is the user (library, generic piece of logic) of a GNSS driver supposed to know what function to call in order to extract this information. Always require a function called `gnss_get_latest_timepulse` to be defined by the application? The "user" doesn't know anything about GPIOs as it is software, not a devicetree node.

**Comment by henrikbrixandersen** [comment]
> Unfortunately the application is not really in a great position to use the 1PPS signal for a few reasons.
> Firstly the PPS signal can stop when the modem is put into sleep mode. The modem driver obviously knows when this happens and can return the documented error, not necessarily true if the application is handling this.

That's a valid point. I didn't consider that.

> Secondly, some modems have weird transient behaviors on the PPS signal at power mode transitions. The application layer cannot possibly handle this in a generic way, while a driver can perform whatever masking operations are necessary to provide only valid values to the application.

Right.

> Thirdly, if delegated to the application, how is the user (library, generic piece of logic) of a GNSS driver supposed to know what function to call in order to extract this information. Always require a function called `gnss_get_latest_timepulse` to be defined by the application? The "user" doesn't know anything about GPIOs as it is software, not a devicetree node.

Not sure I follow, but the first two points are very convincing.

It would be nice to retain the possibility for the application to _also_ get the 1PPS GPIO interrupt for it's own use, while also capturing the timestamp in the GNSS driver, though.

File: include/zephyr/drivers/gnss.h (hunk 1 above)
Comment (JordanYates) [reply include/zephyr/drivers/gnss.h]: done

File: samples/drivers/gnss/src/main.c (hunk 2 above)
Comment (JordanYates) [reply samples/drivers/gnss/src/main.c]: done

**Review by tomi-font — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]
I'm still not entirely convinced, but let's do it :)

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by miguelcarvalhosa** [comment]
Hi everyone. I'm working on an application that must get microsecond-accurate time from a GNSS receiver using the 1PPS signal. I'm using the Quectel LC76G and LC86G receivers.

My question is, i'm not sure how to specify to which GPIO the 1PPS signal is connected.
@JordanYates I see the changes you made to the GNSS sample, but there is no mention on how to feed the 1PPS signal to the SoC.

Any help is appreciated here, thanks in advance!

**Pull request merged**